src/libsync/bulkpropagatorjob.cpp

```diff
@@ -0,0 +1,869 @@
+        uploadParametersData.push_back({std::move(device), oneFile._headers});
+    }
+
+    auto job = std::make_unique<PutMultiFileJob>(propagator()->account(), QUrl{}, std::move(uploadParametersData), this);
+    connect(job.get(), &PutMultiFileJob::finishedSignal, this, &BulkPropagatorJob::slotPutFinished);
+    connect(job.get(), &PutMultiFileJob::uploadProgress, this, &BulkPropagatorJob::slotUploadProgress);
+    //connect(job.get(), &PutMultiFileJob::uploadProgress, device.get(), &UploadDevice::slotJobUploadProgress);
+    connect(job.get(), &QObject::destroyed, this, &BulkPropagatorJob::slotJobDestroyed);
+    //adjustLastJobTimeout(job.get(), oneFile._fileSize);
+    auto jobCopy = job.get();
+    _jobs.append(job.release());
+    jobCopy->start();
+}
+
+void BulkPropagatorJob::slotComputeContentChecksum(SyncFileItemPtr item,
+                                                   UploadFileInfo fileToUpload)
+{
+    qDebug() << "Trying to compute the checksum of the file";
+    qDebug() << "Still trying to understand if this is the local file or the uploaded one";
+    if (propagator()->_abortRequested) {
+        return;
+    }
+
+    const QString filePath = propagator()->fullLocalPath(item->_file);
+
+    // remember the modtime before checksumming to be able to detect a file
+    // change during the checksum calculation - This goes inside of the item->_file
+    // and not the fileToUpload because we are checking the original file, not there
+    // probably temporary one.
+    item->_modtime = FileSystem::getModTime(filePath);
+
+    const QByteArray checksumType = propagator()->account()->capabilities().preferredUploadChecksumType();
+
+    // Maybe the discovery already computed the checksum?
+    // Should I compute the checksum of the original (item->_file)
+    // or the maybe-modified? (fileToUpload._file) ?
+
+    QByteArray existingChecksumType;
+    QByteArray existingChecksum;
+    parseChecksumHeader(item->_checksumHeader, &existingChecksumType, &existingChecksum);
+    if (existingChecksumType == checksumType) {
+        slotComputeTransmissionChecksum(item, fileToUpload, checksumType, existingChecksum);
+        return;
+    }
+
+    // Compute the content checksum.
+    auto computeChecksum = std::make_unique<ComputeChecksum>(this);
+    computeChecksum->setChecksumType(checksumType);
+
+    connect(computeChecksum.get(), &ComputeChecksum::done,
+            this, [this, item, fileToUpload] (const QByteArray &contentChecksumType, const QByteArray &contentChecksum) {
+        slotComputeTransmissionChecksum(item, fileToUpload, contentChecksumType, contentChecksum);
+    });
+    connect(computeChecksum.get(), &ComputeChecksum::done,
+            computeChecksum.get(), &QObject::deleteLater);
+    connect(computeChecksum.get(), &QObject::destroyed, this, [this] (QObject *deletedJob) {
+        _checksumsJobs.erase(std::remove(_checksumsJobs.begin(), _checksumsJobs.end(), deletedJob), _checksumsJobs.end());
```

Comment:
Not sure if that will be a problem, but how big will _checksumJobs usually grow, and how often will this part of the code called? I guess there could happen quite some copying. If it's called often, maybe a linked list would be better, but I currently don't see if random access is more important here.